### PR TITLE
Fallback to action scope key subtypes, if a requested key is missing

### PR DESCRIPTION
**Pull request by mmollaverdi**
For example, given `HttpCall` extends `HttpRequest`, the value for `HttpCall` key will be returned when the `HttpRequest` key is requested (only if a `HttpRequest` key is not set).

See this PR for related prior work:
https://github.com/cashapp/misk/pull/3104

### CODE REVIEW

**Review by jvmakine — APPROVED** [review]

File: misk/src/main/kotlin/misk/web/BoundAction.kt [hunk 1]
```diff
@@ -111,7 +110,6 @@ internal class BoundAction<A : WebAction>(
     val initialSeedData = mapOf<Key<*>, Any?>(
       keyOf<HttpServletRequest>() to request,
       keyOf<HttpCall>() to httpCall,
-      keyOf<HttpRequest>() to httpCall,
```
Comment (adrw) [inline misk/src/main/kotlin/misk/web/BoundAction.kt:114]: If this is set here, does it error now with the changes to ActionScope?

**Review by adrw — APPROVED** [review]
Clever approach, I'm not anticipating any downstream breakages on account of this but would monitor to confirm after this merges.

File: misk/src/main/kotlin/misk/web/BoundAction.kt (hunk 1 above)
Comment (mmollaverdi) [reply misk/src/main/kotlin/misk/web/BoundAction.kt:114]: nope, I'm only falling back to keys for subclasses if the key for the requested type is not set, so if a HttpRequest key is set, that'd still be returned.

File: misk-action-scopes/src/main/kotlin/misk/scope/ActionScope.kt [hunk 2]
```diff
@@ -183,6 +183,20 @@ class ActionScope @Inject internal constructor(
...
+    // Search for any subtype of this key
+    // For example, given `HttpCall` extends `HttpRequest`, the value for `HttpCall` key
+    // will be returned when `HttpRequest` is requested (only if a `HttpRequest` key is not set)
+    val keyClass = key.typeLiteral.rawType
+    val subtypesThreadState = threadState.filterKeys {
+      it.annotation == key.annotation && keyClass.isAssignableFrom(it.typeLiteral.rawType)
+    }
+    if (subtypesThreadState.isNotEmpty()) {
+      val subtypeCachedValue = subtypesThreadState.values.toList()[0]
+      threadState[key] = subtypeCachedValue // cache this value
```
Comment (frojasg) [inline misk-action-scopes/src/main/kotlin/misk/scope/ActionScope.kt:195]: nice!

**Review by frojasg — APPROVED** [review]
LGTM

**Pull request merged**